nuxeo-core/nuxeo-core-test/src/test/java/org/nuxeo/ecm/core/TestGetOrCreateDocument.java

```diff
@@ -0,0 +1,229 @@
+                            TestGetOrCreateDocument.this::computeKey, (CoreSession s, DocumentModel docModel) -> {
+                                DocumentRef ref = docModel.getRef();
+                                if (s.exists(ref)) {
+                                    return s.getDocument(ref);
+                                }
+                                if (latch != null) {
+                                    // Simulate a thread that keeps a document locked
+                                    try {
+                                        latch.await();
+                                    } catch (InterruptedException e) {
+                                        Thread.currentThread().interrupt(); // restore interrupt
+                                        throw new NuxeoException(e);
+                                    }
+                                }
+                                return s.createDocument(docModel);
+                            });
+                } catch (ConcurrentUpdateException e) {
+                    if (latch != null) {
+                        latch.countDown();
+                    }
+                    throw e;
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt(); // restore interrupt
+                    if (latch != null) {
+                        latch.countDown();
+                    }
+                    throw new NuxeoException(e);
+                }
+            });
+        }
+    }
+
+    protected String computeKey(CoreSession session, DocumentModel docModel) {
+        String repositoryName = docModel.getRepositoryName();
+        String parentId = session.getDocument(docModel.getParentRef()).getId();
+        String name = docModel.getName();
+        return repositoryName + "-" + parentId + "-" + name;
+    }
+
+    @Test
+    public void testCreateDocument() throws Exception {
+
+        DocumentModel doc = session.createDocumentModel("/", "file", "File");
+        assertFalse(session.exists(doc.getRef()));
+
+        GetOrCreateDocumentThread t = new GetOrCreateDocumentThread(session, doc);
+        Thread thread = new Thread(t);
+        thread.start();
+        thread.join();
+
+        doc = t.documentModel;
+
+        assertTrue(session.exists(doc.getRef()));
+
+    }
+
+    @Test
+    public void testGetDocument() throws Exception {
+
+        DocumentModel doc = session.createDocumentModel("/", "file", "File");
+        doc = session.createDocument(doc);
+        session.save();
+
+        assertTrue(session.exists(doc.getRef()));
+
+        GetOrCreateDocumentThread t = new GetOrCreateDocumentThread(session, doc);
+        Thread thread = new Thread(t);
+        thread.start();
+        thread.join();
+        doc = t.documentModel;
+
+        assertTrue(session.exists(doc.getRef()));
+
+    }
+
+    @Test
+    public void testLockIsSetWhenCreatingDocument() throws Exception {
```

Comment:
I'd remove this test. It tests the implementation details of getOrCreate but we don't care about the internals in tests; what we care about is the observable behavior.